### PR TITLE
Allow specifying webhook URLs with credentials

### PR DESCRIPTION
**Pull request by michaelbeaumont**
This makes it a bit cleaner to use `office365ConnectorWebhooks` in a `Jenkinsfile`.

### CODE REVIEW

**Comment by damianszczepanik** [comment]
What is the use case you are trying to address?

**Comment by michaelbeaumont** [comment]
I want to use `office365ConnectorWebhooks` in the `options` directive but don't want to put a URL into my `Jenkinsfile`, i.e.:
```
options {
  office365ConnectorWebhooks([
    [name: "Office 365", urlCredentialsId: "teamswebhook", notifyFailure: true]
  ])
}
```

**Comment by damianszczepanik** [comment]
Have you tried this?
```
            withCredentials(bindings: [string(credentialsId: 'OFFICE365_HOOK', variable: 'HOOK_URL')]) {
                office365ConnectorSend webhookUrl: "${HOOK_URL}"
            }
```
This is how I have managed problem with keeping URL out of the project

**Comment by michaelbeaumont** [comment]
I'm not using `office365ConnectorSend`. 

`office365ConnectorWebhooks` is used in `options` and can't access credentials that way.

**Comment by damianszczepanik** [comment]
What is the difference  to have it in option vs in post section?

**Comment by michaelbeaumont** [comment]
Well with `office365ConnectorWebhooks` I can specify all of the conditions under which to send it once, it's kind of the point of `office365ConnectorWebhooks`...

```
[name: "Office 365", urlCredentialsId: "teamswebhook", notifyFailure: true, notifySuccess: true, notifyNotBuilt: true, notifyUnstable: true]
```

`office365ConnectorSend` is just a step. I have to repeat it under each `post` section. And I'm not sure if the message that's sent is the same one.

**Comment by damianszczepanik** [comment]
Well I have single post section so please check if this is enough for you as well. Messages are sent the same for sure

**Comment by michaelbeaumont** [comment]
It's great that you have only one `post` section. But I just explained how it can't be expressed the same way. How can you select for multiple conditions in one post condition? It doesn't seem like you're understanding at all.

`office365ConnectorWebhooks` is much more concise but it's currently unusable with credentials. Your workaround, which, again, uses `office365ConnectorSend` and not `office365ConnectorWebhooks`, leads to worse readability and duplicated code.

**Comment by damianszczepanik** [comment]
Yes, maybe I don't understand the point as what I can do the same by using `always` section as descibed  https://jenkins.io/doc/book/pipeline/syntax/#post

**Comment by michaelbeaumont** [comment]
Yes but `always` is always! Please look at how `office365ConnectorWebhooks` works. It allows me to specify a conjunction of conditions. It is NOT the same as `office365ConnectorSend`.

**Comment by michaelbeaumont** [comment]
@damianszczepanik Do you plan on merging this? I think this really improves **`office365ConnectWebhooks`** with little downside.

**Comment by damianszczepanik** [comment]
> 
> 
> @damianszczepanik Do you plan on merging this? I think this really improves **`office365ConnectWebhooks`** with little downside.

I understand the problem which in fact is. However I don't want to add new responsibility to this plugin as I believe this can be and even should implemented as the part of https://jenkins.io/doc/pipeline/steps/credentials-binding/

**Comment by michaelbeaumont** [comment]
That's a step and can't be used inside `options` so there's no chance it can ever be used in combination with `office365ConnectorWebhooks` but ok, I see this is futile so fork it is then.

**Comment by damianszczepanik** [comment]
> 
> 
> That's a step and can't be used inside `options` so there's no chance it can ever be used in combination with `office365ConnectorWebhooks` but ok, I see this is futile so fork it is then.

My recommendation
1. Check if possible via mentioned plugin and some extension to this one if needed
2. Email jenkinsci-dev mailing groups for help as I believe you might not be the first with such problem. Or maybe this plugin is not designed properly...
if above do not bring the expected result I will review this and merge

**Comment by damon-atkins** [comment]
A web hook in this case is also a password/secret, it should not be stored in git or a pipeline file.

**Comment by damon-atkins** [comment]
And it would be nice if you can configure GUI within the pipeline instead of having to do it in the GUI itself.

The GUI should accept a secret text for the WebUrl Hook instead otherwise this plugin is not secure. Not sure how github notify works or is structured, it might be a good example to follow.   I will say github notify can not be switch off, which is one bad think about it,

PS And support for multibranchpipelines

**Comment by thcp** [comment]
guys, quick note,  i'm using this approach and it works well with azure keyvault:


```
def teamsNotify(STATUS,MESSAGE) {
  script {
    def secrets = [
        [ secretType: 'Secret', name: "teams-token-url", version: '', envVariable: 'HOOK_URL' ],
    ]
    withAzureKeyvault(secrets) {
      office365ConnectorSend message: "${MESSAGE}", status: "${STATUS}", webhookUrl:"${HOOK_URL}"
    }
  }  
}
```

**Comment by damon-atkins** [comment]
Using something like this to override the defaults configured global seems a good idea.
From the example above:
```
options {
  office365ConnectorWebhooks([
    [name: "Office 365",  notifyFailure: true]
  ])
}
```
Any one have code to configure the defaults in init.groovy or does this plugin work with the configuration-as-code-plugin.

File: src/main/java/jenkins/plugins/office365connector/Webhook.java [hunk 1]
```diff
@@ -36,6 +40,7 @@
 
     private String name;
     private String url;
+    private String urlCredentialsId;
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Webhook.java]: then add documentation what is the purpose of having first and the second url

File: src/main/java/jenkins/plugins/office365connector/Webhook.java [hunk 2]
```diff
@@ -51,12 +56,31 @@
     private List<Macro> macros = Collections.emptyList();
 
     @DataBoundConstructor
-    public Webhook(String url) {
-        this.url = url;
+    public Webhook() {
+    }
+
+    public String getUrl(Run run) {
+        if (this.url != null) {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Webhook.java]: not very good about strings - how about "" ?

File: src/main/java/jenkins/plugins/office365connector/utils/FormUtils.java [hunk 3]
```diff
@@ -17,17 +17,21 @@
      * @param value user input to validate
      * @return OK or Error on pending user input
      */
-    public static FormValidation formValidateUrl(String value) {
+    public static FormValidation formValidateUrl(String value, String urlCredentialsId) {
+        if (StringUtils.isBlank(value)) {
+            if (!StringUtils.isBlank(urlCredentialsId)) {
+                return FormValidation.ok();
+            } else {
+                return FormValidation.error("Either URL or URL Credentials must be non-empty");
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/utils/FormUtils.java]: anyway, did you try to add support for credentials using plugin which I mentioned earlier?

if so the only IMO thing which should be changed is https://github.com/jenkinsci/office-365-connector-plugin/pull/102/files#diff-bd70be8a13679d8af35645c64df15ceaR7 from textbox to password

File: src/main/resources/jenkins/plugins/office365connector/Webhook/help-urlCredentialsId.html [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<div align="help">
+    Allows specifying the URL with credentials.
```
Comment (damianszczepanik) [inline src/main/resources/jenkins/plugins/office365connector/Webhook/help-urlCredentialsId.html]: how?

File: src/main/java/jenkins/plugins/office365connector/Webhook.java (hunk 2 above)
Comment (michaelbeaumont) [reply src/main/java/jenkins/plugins/office365connector/Webhook.java]: `url` and `urlCredentialsId` are always set using `fixEmptyAndTrim` so `""` is normalized to `null`.

https://github.com/jenkinsci/office-365-connector-plugin/blob/2560636e8c2ff8e56a7a27beadbd8823084fd883/src/main/java/jenkins/plugins/office365connector/Webhook.java#L77-L79

File: src/main/resources/jenkins/plugins/office365connector/Webhook/help-urlCredentialsId.html (hunk 4 above)
Comment (michaelbeaumont) [reply src/main/resources/jenkins/plugins/office365connector/Webhook/help-urlCredentialsId.html]: Made it clear it's using the `id`

File: src/main/java/jenkins/plugins/office365connector/Webhook.java (hunk 1 above)
Comment (michaelbeaumont) [reply src/main/java/jenkins/plugins/office365connector/Webhook.java]: Added a comment here in the code

File: src/main/java/jenkins/plugins/office365connector/utils/FormUtils.java (hunk 3 above)
Comment (michaelbeaumont) [reply src/main/java/jenkins/plugins/office365connector/utils/FormUtils.java]: It's supposed to contain the id of a credential, does that need to be hidden using `password`?

**Comment by damianszczepanik** [comment]
I'm OK with that change. Can you update README.md in a meantime and also fix https://app.codacy.com/app/damianszczepanik/office-365-connector-plugin/pullRequest?prid=3317103 ?
Give me some time to test it first but I want to include this patch in next version.

Also as the https://github.com/jenkinsci/office-365-connector-plugin/pull/116/commits/b19900d3898b20e13603741cab8a7532afa96966 is merged I consider to release new major version so compatibility is not critical (but will be helpful for sure). And if so, maybe simple change from textarea to password would be enough and the code would be much simpler

**Comment by damianszczepanik** [comment]
@michaelbeaumont have you tried this https://groups.google.com/forum/?nomobile=true#!searchin/jenkinsci-dev/secret|sort:date/jenkinsci-dev/Cm02PZVeIAQ/SlHSvG1mCQAJ or https://groups.google.com/forum/?nomobile=true#!searchin/jenkinsci-dev/secret|sort:date/jenkinsci-dev/gv8gyJ4XKmk/lXFi5f1CBAAJ ?

**Comment by michaelbeaumont** [comment]
> @michaelbeaumont have you tried this [https://groups.google.com/forum/?nomobile=true#!searchin/jenkinsci-dev/secret|sort:date/jenkinsci-dev/Cm02PZVeIAQ/SlHSvG1mCQAJ](https://groups.google.com/forum/?nomobile=true#!searchin/jenkinsci-dev/secret%7Csort:date/jenkinsci-dev/Cm02PZVeIAQ/SlHSvG1mCQAJ) or [https://groups.google.com/forum/?nomobile=true#!searchin/jenkinsci-dev/secret|sort:date/jenkinsci-dev/gv8gyJ4XKmk/lXFi5f1CBAAJ](https://groups.google.com/forum/?nomobile=true#!searchin/jenkinsci-dev/secret%7Csort:date/jenkinsci-dev/gv8gyJ4XKmk/lXFi5f1CBAAJ) ?

Sorry, I'm confused, what do those topics have to do with this PR?

**Comment by damianszczepanik** [comment]
They provide information how to convert plain text field into secret one. So this is what you are trying to do now but instead of having two separate fields which work exclusively it's better to convert existing one into secret version

**Comment by damianszczepanik** [comment]
@michaelbeaumont should I want for this before next release?

**Comment by damon-atkins** [comment]
Maybe if the url starts with https:// its a url otherwise its a secret id.

File: src/main/java/jenkins/plugins/office365connector/Webhook.java [hunk 5]
```diff
@@ -34,7 +38,9 @@
     public static final Integer DEFAULT_TIMEOUT = 30000;
 
     private String name;
+    // The url can be specified using either a string or the id of plain text credentials
     private String url;
+    private String urlCredentialsId;
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Webhook.java]: Did you look into references documentation to remove second filed for url and to have only the old one?

**Comment by damianszczepanik** [comment]
@michaelbeaumont are you going to continue work on this?

**Comment by michaelbeaumont** [comment]
@damianszczepanik i don't think so. we're moving to a different workflow to notify through teams so the plugin isn't necessary for us anymore. i also don't really understand the change you were requesting or how having a password field instead of plain text helps anything.

**Comment by fl-max** [comment]
Bummer that this wasn't implemented. Would have made `office365ConnectorWebhooks` in `options{ }` actually useful...

For anyone stumbling upon this thread, do **NOT** store the connector url in your Jenkinsfile. It should be treated as a password.

**Comment by justinmchase** [comment]
@michaelbeaumont re:

> we're moving to a different workflow to notify through teams

Can you elaborate on this? My company won't allow us to use plain webhooks so if there was an alternate technique to get messages into teams in a more secure manner I'd be interested in hearing about it.

**Comment by amit-k-yadav** [comment]
I am sure most of you already know this.

But for anyone who wants to use `office365ConnectorWebhooks` and avoid pasting the actual webhook in `url` and no idea how to do it, here's what they can do:

1. In Jenkins UI, go to 'Dashboard' -> 'Manage Jenkins' -> 'Configure System' and search for '**_Office 365 Connector_**'
   
Screenshot 2022-01-13 at 5 56 56 PM

Screenshot 2022-01-13 at 6 16 24 PM

2. Enter Microsoft Teams' WebHook URL and a name
3. Add the below code in your Jenkinsfile:

    ```groovy
    options {
        office365ConnectorWebhooks([[
            startNotification: true,
            notifySuccess: true
        ]])
    }
    ```

**Comment by amit-k-yadav** [comment]
I am sure most of you already know this.

But for anyone who wants to use `office365ConnectorWebhooks` and avoid pasting the actual webhook in `url` and no idea how to do it, here's what they can do:

1. In Jenkins UI, go to 'Dashboard' -> 'Manage Jenkins' -> 'Configure System' and look for '**_Office 365 Connector_**'
<hr/>
Screenshot 2022-01-13 at 5 56 56 PM
<hr/>
Screenshot 2022-01-13 at 6 16 24 PM
<hr/>

2. Enter Microsoft Teams' WebHook URL and a name
3. Add the below code in your Jenkinsfile:

    ```groovy
    options {
        office365ConnectorWebhooks([[
            startNotification: true,
            notifySuccess: true
        ]])
    }
    ```

The values in the global settings will be used as default when using the plugin for a job. These settings can then be overriden in the job.

For more details -> https://github.com/jenkinsci/office-365-connector-plugin/blob/master/README.md#jenkins-global-configuration

**Comment by rounak-codiant** [comment]
> I am sure most of you already know this.
> 
> But for anyone who wants to use `office365ConnectorWebhooks` and avoid pasting the actual webhook in `url` and no idea how to do it, here's what they can do:
> 
> 1. In Jenkins UI, go to 'Dashboard' -> 'Manage Jenkins' -> 'Configure System'
>     Screenshot 2022-01-13 at 5 56 56 PM
> 2. Look for '**_Office 365 Connector_**' and Enter Microsoft Teams WebHook URL and name
>     Screenshot 2022-01-13 at 6 16 24 PM
> 3. Add the below code in your Jenkinsfile:
>    ```groovy
>    options {
>        office365ConnectorWebhooks([[
>            startNotification: true,
>            notifySuccess: true
>        ]])
>    }
>    ```
> 
> The values in the global settings will be used as default when using the plugin for a job. These settings can then be overriden in the job.
> 
> For more details -> https://github.com/jenkinsci/office-365-connector-plugin/blob/master/README.md#jenkins-global-configuration

What If you have multiple URLs for multiple channels

**Comment by amit-zact** [comment]
@rounak-codiant you can override the global URL inside your pipelines like below:

```groovy
job('Example Job Name') {
    description 'Example description'
    properties {
        office365ConnectorWebhooks {
            webhooks {
                webhook {
                    name('Example Webhook Name')
                    url('https://outlook.office.com/webhook/123456...')
                    startNotification(false)
                    notifySuccess(true)
                    notifyAborted(false)
                    notifyNotBuilt(false)
                    notifyUnstable(true)
                    notifyFailure(true)
                    notifyBackToNormal(true)
                    notifyRepeatedFailure(false)
                    timeout(30000)
                }
            }
        }
    }

    // Webhook Macro Configuration
    configure {
        // Example: Conditioning webhook trigger on build parameter 'version' being equal to 'latest'
        // Templates are defined as token macros https://github.com/jenkinsci/token-macro-plugin
        it / 'properties' / 'jenkins.plugins.office365connector.WebhookJobProperty' / 'webhooks' / 'jenkins.plugins.office365connector.Webhook' / 'macros' << 'jenkins.plugins.office365connector.model.Macro' {
          template('${ENV, var="version"}')
          value('latest')
        }
    }
}
```
See [these docs](https://github.com/jenkinsci/office-365-connector-plugin/blob/master/README.md#dsl) for more details.

**Comment by amit-k-yadav** [comment]
@rounak-codiant you can override the global URL inside your pipelines like below:

```groovy
job('Example Job Name') {
    description 'Example description'
    properties {
        office365ConnectorWebhooks {
            webhooks {
                webhook {
                    name('Example Webhook Name')
                    url('https://outlook.office.com/webhook/123456...')
                    startNotification(false)
                    notifySuccess(true)
                    notifyAborted(false)
                    notifyNotBuilt(false)
                    notifyUnstable(true)
                    notifyFailure(true)
                    notifyBackToNormal(true)
                    notifyRepeatedFailure(false)
                    timeout(30000)
                }
            }
        }
    }

    // Webhook Macro Configuration
    configure {
        // Example: Conditioning webhook trigger on build parameter 'version' being equal to 'latest'
        // Templates are defined as token macros https://github.com/jenkinsci/token-macro-plugin
        it / 'properties' / 'jenkins.plugins.office365connector.WebhookJobProperty' / 'webhooks' / 'jenkins.plugins.office365connector.Webhook' / 'macros' << 'jenkins.plugins.office365connector.model.Macro' {
          template('${ENV, var="version"}')
          value('latest')
        }
    }
}
```
See [these docs](https://github.com/jenkinsci/office-365-connector-plugin/blob/master/README.md#dsl) for more details.

**Comment by faandg** [comment]
This is a funny, and kind of sad, PR.
To summarize;

`office365ConnectorWebhooks` (in options block) != `office365ConnectorSend` (which can be specified anywhere)
`office365ConnectorWebhooks` has more fine grained configuration options on when to notify.

**why not in stage?**
Stages can run in parallel or can fail or never be reached unless you've prepared for that eventuality, hence this may lead to never receiving a notification at all.

**why not scripted (in stage)**
on top the above, possibly you could introduce a bunch of scripted pipeline code with `office365ConnectorSend` to achieve the same thing, but declarative syntax is always a better choice where possible, especially for logic everybody uses like sending notifications.

**why not in post?**
post's subdivides are `always, changed, fixed, regression, aborted, failure, success, unstable, unsuccessful, cleanup` and they cannot map on the plugin's `notifySuccess, notifyAborted, notifyNotBuilt, notifyUnstable, notifyFailure, notifyBackToNormal, notifyRepeatedFailure`, you would need to define the webhook for each post subdivide or use `always`. In order to reduce the spam of notifications sent by `always` you would need to introduce custom logic again.
In other words post is not ideal for something you already want to use a similar `case switch` for.

**why not in global settings?**
not an option when you have several teams with different webhook URLs and you end up overriding it in the pipeline anyway where it is again visible (because the essence of the story is we don't want it to show for security purposes, we want to handle it like a secret).

**Using `Credentials Binding` in the `Options` block** seems like the best approach, but as I understand falls out of the scope of this plugin and is not allowed by the declarative syntax, so I filed an issue to see if it can be allowed https://issues.jenkins.io/browse/JENKINS-70355

Either way it might also be an improvement to align the configuration options of `office365ConnectorWebhooks` (in options block) and `office365ConnectorSend` so you can use either one for the same purpose, if possible.

**Comment by michaelbeaumont** [comment]
Finally giving up on this one 😬

**Pull request closed without merging**